Add documentation and scripts on running e2e test locally

nawazkh opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
- This PR adds documentation on running e2e tests locally, both in a local linux/unix machine and on a dev VM. 
- Adds a `peer_vnets.sh` script that the user can use to 
  - peer vnets
  - create private dns zone
  - unblock required ports on Azure NSGs
- Adds instructions on adding required kustomize workarounds to get a flavor tested locally. 


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5268 

**Special notes for your reviewer**:
<!-- Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/developers/releasing.md#release-support for more information about which changes are eligible for backport -->


**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [x] includes documentation
- [ ] adds unit tests
- [ ] cherry-pick candidate

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add documentation on running e2e tests locally
```

nawazkh commented:
I was able to get one test run locally
```
.
.
.
    STEP: PASSED! @ 01/23/25 17:17:53.934
  Jan 23 17:17:53.934: INFO: Cleaning up after "Workload cluster creation Local test: Creating a self-managed VM based cluster using API Server ILB feature gate and fully spec-ed out APIServer ILB template [OPTIONAL][API-Server-ILB] with three controlplane node and three worker nodes" spec
  Jan 23 17:17:53.934: INFO: Dumping all the Cluster API resources in the "capz-e2e-wp84yi" namespace
  STEP: Redacting sensitive information from logs @ 01/23/25 17:18:04.938
  Jan 23 17:18:05.022: WARNING: Redact logs command failed: exit status 1
  INFO: "with three controlplane node and three worker nodes" ran for 14m40s on Ginkgo node 1 of 10 and reported junit test to file /Users/nawazhussain/msftcode/cluster-api-provider-azure/_artifacts/test_e2e_junit.e2e_suite.1.xml
  << Timeline
------------------------------
[SynchronizedAfterSuite] PASSED [0.000 seconds]
[SynchronizedAfterSuite]
/Users/nawazhussain/msftcode/cluster-api-provider-azure/test/e2e/e2e_suite_test.go:123

  Timeline >>
  STEP: Tearing down the management cluster @ 01/23/25 17:18:05.024
  << Timeline
------------------------------
[ReportAfterSuite] PASSED [0.025 seconds]
[ReportAfterSuite] Autogenerated ReportAfterSuite for --junit-report
autogenerated by Ginkgo
------------------------------

Ran 1 of 33 Specs in 1106.852 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 32 Skipped


Ginkgo ran 1 suite in 18m58.0059155s
Test Suite Passed
/Library/Developer/CommandLineTools/usr/bin/make clean-release-git
git restore ./*manager_image_patch.yaml ./*manager_pull_policy.yaml
================ REDACTING LOGS ================
All sensitive variables are redacted
```

For my reference, the chain of options used to run an e2e test is
`GINKGO_FOCUS="Local test: Creating a self-managed VM based cluster using API Server ILB feature gate and fully spec-ed out APIServer ILB template" USE_LOCAL_KIND_REGISTRY=false SKIP_CLEANUP="true" SKIP_LOG_COLLECTION="true" REGISTRY=nhkregistry.azurecr.io MGMT_CLUSTER_TYPE="aks" EXP_APISERVER_ILB=true scripts/ci-e2e.sh`

nawazkh commented:
One more success; I have figured out the formula for running e2e tests locally for non-aks flavors. Will wire up a Readme explaining it. 
```
GINKGO_FOCUS="With 3 control-plane nodes and 2 Linux and 2 Windows worker nodes" USE_LOCAL_KIND_REGISTRY=false SKIP_CLEANUP="true" SKIP_LOG_COLLECTION="true" REGISTRY=nhkregistry.azurecr.io MGMT_CLUSTER_TYPE="aks" EXP_APISERVER_ILB=true AZURE_LOCATION="westeurope" scripts/ci-e2e.sh
```

```bash
------------------------------
[SynchronizedAfterSuite] PASSED [0.000 seconds]
[SynchronizedAfterSuite]
/Users/nawazhussain/msftcode/cluster-api-provider-azure/test/e2e/e2e_suite_test.go:123

  Timeline >>
  STEP: Tearing down the management cluster @ 03/06/25 00:07:12.08
  << Timeline
------------------------------
[ReportAfterSuite] PASSED [0.007 seconds]
[ReportAfterSuite] Autogenerated ReportAfterSuite for --junit-report
autogenerated by Ginkgo
------------------------------

Ran 1 of 31 Specs in 2405.318 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 30 Skipped
```

nawazkh commented:
/test pull-cluster-api-provider-azure-windows-with-ci-artifacts

nawazkh commented:
/retest

nawazkh commented:
/retest

nawazkh commented:
/retest

nawazkh commented:
/retest

nawazkh commented:
/retest

nawazkh commented:
/retest

nawazkh commented:
Failing `pull-cluster-api-provider-azure-windows-with-ci-artifacts` and `pull-cluster-api-provider-azure-windows-custom-builds` make me wonder if we should pin these tests to a stable K8s version when run in PRs
And have a periodic version of these tests running in the CNCF sub ?

nawazkh commented:
@willie-yao @Jont828 , the above PR is ready for review. We can ignore `pull-cluster-api-provider-azure-windows-with-ci-artifacts` and `pull-cluster-api-provider-azure-windows-custom-builds` test signals. Reference test-infra PR: https://github.com/kubernetes/test-infra/pull/34554

Please take a look at this PR. We would want to ensure that the make targets are intact and do not break any other testings.

willie-yao commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+   - `/spec/template/spec/preKubeadmCommands/0`
+      - Add prekubeadm command updating the `/etc/hosts` of the worker node of the type windows.
+
+A Sample kustomize command updating a prow template via its kustomization.yaml is pasted below.
```
```suggestion
A sample kustomize command for updating a prow template via its kustomization.yaml is pasted below.
```

nawazkh replied:
addressed via 0c7bb63216b299c688bb945af39ee083c4775638

willie-yao commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+
+#### Peer Vnets of the management cluster and the workload cluster
+
+Peering VNets, creation of private DNS zone with the FQDN of the workload cluster and updating of NSGs of management and workload cluster can be achieved by `peer-vnets.sh` present in `scripts/peer-vnets.sh`
```
```suggestion
Peering VNets, creation of private DNS zone with the FQDN of the workload cluster and updating of NSGs of management and workload cluster can be achieved by `scripts/peer-vnets.sh`
```

nawazkh replied:
addressed via 0c7bb63216b299c688bb945af39ee083c4775638

willie-yao commented on scripts/peer-vnets.sh:
```diff
@@ -0,0 +1,426 @@
+wait_and_fix_nsg_rules() {
+    local tcp_ports="443 5986 6443"
+    local udp_ports="53 123"
+    local timeout=3000     # seconds to wait per NSG for the appearance of an NRMS-Rule-101 rule
```
Is this timeout a bit excessive? I know the rule takes a while to be populated but not sure if 50 minutes is too much. This is probably not a big issue though since I don't think we'll ever wait that long, and the rule always gets populated by some Azure job.

nawazkh replied:
It is definitely on the higher end. But in my testing, a 3 control plane - 2 worker node cluster takes about 30 mins to come up. And the NSG rules can take some more time. 
`3000` seemed like a good number.

willie-yao reviewed (commented):
Wow nice work here Nawaz! This is super helpful for unblocking us. Honestly I think you can retitle this PR, as there is much more work here than just documentation. Also left a few comments that are mostly nits.

nawazkh commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+- The below steps for self-managed templates only. Does not apply to AKS workload clusters.
+
+- The below steps are for running the tests in a intuned device. If you are going to run the local tests from a dev machine in Azure, you have to first follow below steps and then proceed ahead.
+  1. Create a managed Identity
```
addressed via 0c7bb63216b299c688bb945af39ee083c4775638

nawazkh commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+
+- The below steps for self-managed templates only. Does not apply to AKS workload clusters.
+
+- The below steps are for running the tests in a intuned device. If you are going to run the local tests from a dev machine in Azure, you have to first follow below steps and then proceed ahead.
```
addressed via 0c7bb63216b299c688bb945af39ee083c4775638

nawazkh commented:
> Wow nice work here Nawaz! This is super helpful for unblocking us. Honestly I think you can retitle this PR, as there is much more work here than just documentation. Also left a few comments that are mostly nits.

Thank you for the review Willie, please take a look again! Addressed your comments!

nawazkh commented:
We have marked `pull-cluster-api-provider-azure-windows-custom-builds` and `pull-cluster-api-provider-azure-windows-with-ci-artifacts` as optional and created periodic jobs instead since they pull latest k/k and might not convey the right signal.

willie-yao commented on templates/flavors/clusterclass/kubeadm-controlplane-template.yaml:
```diff
@@ -61,4 +61,4 @@ spec:
         postKubeadmCommands: []
         preKubeadmCommands: []
-        verbosity: 10
+        verbosity: 10
```
nit: newline at end of file

willie-yao reviewed (commented):
quick nit, otherwise lgtm!

jackfrancis commented:
Should we change the title as this is more than documentation? (it looks like it's actually bringing changes that make it possible to run E2E locally in certain environments)

jackfrancis commented:
/test ls

mboersma commented on Makefile:
```diff
@@ -726,6 +736,24 @@ test-cover: test ## Run tests with code coverage and generate reports.
+create-bootstrap: $(KUBECTL) ## Create bootstrap cluster (AKS or KIND) for CAPZ testing. Default is KIND.
+	@echo "Creating bootstrap cluster with type: $(MGMT_CLUSTER_TYPE)"
+	@if [ "$(MGMT_CLUSTER_TYPE)" == "aks" ]; then \
+		if [ -z "$(AZURE_SUBSCRIPTION_ID)" ]; then \
```
Should some of this variable checking and defaulting be left to `./scripts/aks-as-mgmt.sh`? It's a big chunk of shell code in the Makefile.

nawazkh replied:
Moving the defaulting part to `aks-as-mgmt.sh` script.

mboersma commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+
+### Running e2e tests locally using API Server ILB's networking solution
+
+Building upon the [challenges and solutions](#challenges-and-solutions) from above, running an e2e test locally in a restricted environment calls from some workarounds to the prow templates and the e2e test itself.
```
```suggestion
Building upon the [challenges and solutions](#challenges-and-solutions) from above, running an e2e test locally in a restricted environment calls for some workarounds in the prow templates and the e2e test itself.
```

mboersma commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+- The below steps for self-managed templates only. Does not apply to AKS workload clusters.
+
+- The below steps are for running the tests in an intuned device. If you are going to run the local tests from a dev machine in Azure, you have to first follow below steps and then proceed ahead.
+  1. Create a user assigned managed identity
```
```suggestion
  1. Create a user-assigned managed identity
```

mboersma commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+1. AzureCluster
+   - `/spec/networkSpec/apiServerLB`
+      - Add FrontendIP
+      - Add an associated private IP to be leveraged by an intern ILB
```
```suggestion
      - Add an associated private IP to be leveraged by an internal ILB
```

mboersma commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+
+#### Peer Vnets of the management cluster and the workload cluster
+
+Peering VNets, creation of private DNS zone with the FQDN of the workload cluster and updating of NSGs of management and workload cluster can be achieved by `scripts/peer-vnets.sh`
```
```suggestion
Peering VNets, creating a private DNS zone with the FQDN of the workload cluster, and updating NSGs of the management and workload clusters can be achieved by running `scripts/peer-vnets.sh`.
```

mboersma reviewed (commented):
lgtm overall, just had some suggestions for better wording.

jackfrancis commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-e2e-workload-upgrade

mboersma commented:
/hold

Until comments are addressed.

nawazkh commented on docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md:
```diff
@@ -124,3 +124,154 @@ Once the tilt UI is up and running
+
+- The below steps for self-managed templates only. Does not apply to AKS workload clusters.
+
+- The below steps are for running the tests in an intuned device. If you are going to run the local tests from a dev machine in Azure, you have to first follow below steps and then proceed ahead.
```
Sorry for the wrong wording. The below steps are only required when running the e2e tests from an Dev machine (in Azure).

mboersma reviewed (approved):
/lgtm
/hold cancel

Great work here!

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-e2e-workload-upgrade

nawazkh commented:
/retest

nawazkh commented:
/test pull-cluster-api-provider-azure-ci-entrypoint

willie-yao reviewed (approved):
/lgtm

willie-yao commented:
/approve
